clang/tools/clang-offload-deps/ClangOffloadDeps.cpp

```diff
@@ -0,0 +1,254 @@
+//===----------- clang-offload-deps/ClangOffloadDeps.cpp ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Implementation of the clang-offload-deps tool. This tool is intended to be
+/// used by the clang driver for offload linking with static offload libraries.
+/// It takes linked host image as input and produces bitcode files, one per
+/// offload target, containing references to symbols that must be defined in the
+/// target images. Dependence bitcode file is then expected to be compiled to an
+/// object by the driver using the appropriate offload target toolchain, and
+/// dependence object added to the target linker as input together with the
+/// other inputs. References to the symbols in dependence object should ensure
+/// that target linker pulls in necessary symbol definitions from the input
+/// static libraries.
+///
+//===----------------------------------------------------------------------===//
+
+#include "clang/Basic/Version.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Triple.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#ifndef NDEBUG
+#include "llvm/IR/Verifier.h"
+#endif // NDEBUG
+#include "llvm/Object/ObjectFile.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/ErrorOr.h"
+#include "llvm/Support/Signals.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/WithColor.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define SYMBOLS_SECTION_NAME ".tgtsym"
+
+using namespace llvm;
+using namespace llvm::object;
+
+static cl::opt<bool> Help("h", cl::desc("Alias for -help"), cl::Hidden);
+
+// Mark all our options with this category, everything else (except for -version
+// and -help) will be hidden.
+static cl::OptionCategory
+    ClangOffloadDepsCategory("clang-offload-deps options");
+
+static cl::list<std::string> Outputs("outputs", cl::CommaSeparated,
```

Comment:
Usually multiple outputs within the driver are handled either via file lists (TY_Tempfilelist) or file tables (TY_Tempfiletable).  Driver implementation would easier and more inline with other multi-output tools if this tool could generate a file list listing all the outputs separated by \n.
@mdtoguchi - what do you think?

Reply:
I would expect that the outputs match the targets, which is in-line with the bundler.  I'm OK with how things are.